service/history/queues/monitor.go

```diff
@@ -115,34 +170,44 @@ func (m *monitorImpl) SetReaderWatermark(readerID int32, watermark tasks.Key) {
 	watermark.FireTime = watermark.FireTime.Truncate(monitorWatermarkPrecision)
 	watermark.TaskID = 0
 
-	progress := m.progressByReader[readerID]
-	if progress == nil {
-		m.progressByReader[readerID] = &readerProgess{
+	stats := m.readerStats[readerID]
+	if stats.progress.watermark.CompareTo(watermark) != 0 {
+		stats.progress = readerProgess{
 			watermark: watermark,
 			attempts:  1,
 		}
-		return
-	}
-	if progress.watermark.CompareTo(watermark) != 0 {
-		progress.watermark = watermark
-		progress.attempts = 1
+		m.readerStats[readerID] = stats
```

Comment:
any reason why not store pointer in map?

Reply:
I tried to avoid alloc objects on heap if it's easy to do. In this case, the struct is quite small, so the cost of copying the value is negligible. 

I understand it makes the map assignment a bit awkward, but it also means I don't need to worry about potential npe issues.